### PR TITLE
[SPARK-38970][CI] Skip build-and-test workflow on forks when scheduled

### PR DESCRIPTION
**Pull request by EnricoMi**
### What changes were proposed in this pull request?
The `configure-jobs` and `precondition` jobs in the `build-and-test` workflow should be skipped on fork repositories as all subsequent / dependent jobs are already skipped.

### Why are the changes needed?
The `build and test` workflow is scheduled to run six times a day. The jobs then only run for `type == 'scheduled'` inside apache/spark repository, not for forks. However, the `configure-jobs` and `precondition` jobs always run, even though all subsequent jobs are skipped on forks.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
No

### CODE REVIEW

**Review by dcoliversun — APPROVED** [review]
Good job. Could you please use jira and rename PR name as [SPARK-XXX][CI]?

**Review by HyukjinKwon — COMMENTED** [review]
I think it's fine .. but would be great to have @dongjoon-hyun or @Yikun 's look here.

**Review by Yikun — COMMENTED** [review]
Besides comments inline, a question: should we also remove scheduler job skip logic in [`build`](https://github.com/apache/spark/blob/a2ac450896fd089bdbdf5447a336228fe9f8b879/.github/workflows/build_and_test.yml#L150) and [`pyspark`](https://github.com/apache/spark/blob/a2ac450896fd089bdbdf5447a336228fe9f8b879/.github/workflows/build_and_test.yml#L295-L296) after we added skip logic in `precondition` because these two job are depends on `precondition`.

**Comment by Yikun** [comment]
BTW, this is only skip precondtion (Check changes) but `configure-jobs` is still running in fork repo:

image

**Comment by EnricoMi** [comment]
You are right, moving the logic into `configure-jobs`, on which all other jobs depend on, allows to disable the entire workflow for `scheduled` events on forks and simplify logic in dependent jobs.

**Comment by EnricoMi** [comment]
@Yikun The expression `github.event.type` does not exist for scheduled events. The only content of `event` is the `schedule` field.

In fact, the payload for that event is not even documented: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads

But there is `github.event_name`: https://docs.github.com/en/actions/learn-github-actions/contexts#github-context

**Comment by EnricoMi** [comment]
@Yikun @HyukjinKwon I have simplified logic as suggested

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -140,11 +145,10 @@ jobs:
   build:
     name: "Build modules (${{ format('{0}, {1} job', needs.configure-jobs.outputs.branch, needs.configure-jobs.outputs.type) }}): ${{ matrix.modules }} ${{ matrix.comment }} (JDK ${{ matrix.java }}, ${{ matrix.hadoop }}, ${{ matrix.hive }})"
     needs: [configure-jobs, precondition]
-    # Run scheduled jobs for Apache Spark only
-    # Run regular jobs for commit in both Apache Spark and forked repository
+    # Run for all scheduled jobs
+    # Run for all other jobs only when changes exist
```
Comment (Yikun) [inline .github/workflows/build_and_test.yml:149]: I think the original note still valid in here, and more useful. WDYT?

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -284,12 +288,12 @@ jobs:
 
   pyspark:
     needs: [configure-jobs, precondition]
-    # Run PySpark coverage scheduled jobs for Apache Spark only
-    # Run scheduled jobs with JDK 17 in Apache Spark
-    # Run regular jobs for commit in both Apache Spark and forked repository
+    # Run PySpark coverage scheduled jobs
+    # Run scheduled jobs only with JDK 17
+    # Run regular jobs only if pyspark changes exist
```
Comment (Yikun) [inline .github/workflows/build_and_test.yml:293]: same

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -393,9 +397,11 @@ jobs:
 
   sparkr:
     needs: [configure-jobs, precondition]
+    # Run scheduled jobs only with JDK 17
+    # Run regular jobs only if sparkr changes exist
```
Comment (Yikun) [inline .github/workflows/build_and_test.yml:401]: better also mentioned for `Apache Spark only` and `for commit in both Apache Spark and forked repository`

File: .github/workflows/build_and_test.yml [hunk 4]
```diff
@@ -576,6 +582,7 @@ jobs:
 
   java-11-17:
     needs: [configure-jobs, precondition]
+    # Run regular jobs only if changes exist
```
Comment (Yikun) [inline .github/workflows/build_and_test.yml:585]: better also mentioned for `commit in both Apache Spark and forked repository`

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -48,6 +48,11 @@ jobs:
   configure-jobs:
     name: Configure jobs
     runs-on: ubuntu-20.04
+    # All other jobs in this workflow depend on this job,
+    # so the entire workflow is skipped when these conditions evaluate to false:
+    # Run all jobs for Apache Spark repository
+    # Run only non-scheduled jobs for forked repositories
```
Comment (Yikun) [inline .github/workflows/build_and_test.yml:54]: Good notes, but we better to still keep below notes to mention fork/upstream trigger scopre

**Review by Yikun — APPROVED** [review]
LGTM, some nits I think it could be better.

Could also have a local test on your repo? And comments the link in here?

It's easy to test this by setting the default branch to your current branch (branch-dont-check-changes-on-forks) in your repo ([1] or [2]).

[1] https://github.com/EnricoMi/spark/settings/branches
[2] https://github.com/G-Research/spark/settings/branches

**Comment by EnricoMi** [comment]
> Could also have a local test on your repo? And comments the link in here?

Good idea! I have changed the schedule for all six crons and tested them here: https://github.com/EnricoMi/spark/actions/workflows/build_and_test.yml?query=event%3Aschedule

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (EnricoMi) [reply .github/workflows/build_and_test.yml]: I have reverted and amended the original comment.

File: .github/workflows/build_and_test.yml (hunk 5 above)
Comment (EnricoMi) [reply .github/workflows/build_and_test.yml:54]: I have adjusted the comments and the one condition.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**